lib/model/store.dart

```diff
@@ -673,6 +673,128 @@ class PerAccountStore extends PerAccountStoreBase with
 
   final AutocompleteViewManager autocompleteViewManager = AutocompleteViewManager();
 
+  /// Whether the user has permission to delete the message, as of [byDate].
+  ///
+  /// For a value of [byDate], use [ZulipBinding.instance.utcNow].
+  // There might be a better place for this to live…but it uses
+  // message, channel, self-user, and user-group data, so seemed simplest to
+  // put it here where all that is available.
+  bool selfCanDeleteMessage(int messageId, {required DateTime byDate}) {
+    // Compare web's message_delete.get_deletability.
+
+    final message = messages[messageId];
+    if (message == null) return false; // TODO(log)
+
+    final ZulipStream? channel;
+    if (message is StreamMessage) {
+      channel = streams[message.streamId];
+      // TODO(log) if channel null here?
+    } else {
+      channel = null;
+    }
+
+    if (channel != null && channel.isArchived == true) {
+      return false;
+    }
+
+    if (realmCanDeleteAnyMessageGroup != null
+        && selfHasPermissionForGroupSetting(realmCanDeleteAnyMessageGroup!,
+             GroupSettingType.realm, 'can_delete_any_message_group')) {
+      // realmCanDeleteAnyMessageGroup always present as of FL 281.
+      // TODO(server-10) make required, simplify, delete this FL-281 comment
+      return true;
+    }
+
+    if (channel != null) {
+      if (channel.canDeleteAnyMessageGroup != null
+          && selfHasPermissionForGroupSetting(channel.canDeleteAnyMessageGroup!,
+               GroupSettingType.stream, 'can_delete_any_message_group')) {
+        // channel.canDeleteAnyMessageGroup always present as of FL 407.
+        // TODO(server-11) make required, simplify, delete this FL-407 comment
+        return true;
+      }
+    }
+
+    final sender = getUser(message.senderId);
+    if (sender == null) return false;
+
+    if (
+      sender.userId != selfUserId
+      && !(sender.isBot && sender.botOwnerId == selfUserId)
+    ) {
+      return false;
+    }
+
+    // Web returns false here for local-echoed message objects;
+    // that's impossible here because `message` can't be an [OutboxMessage]
+    // (it's a [Message] from [MessageStore.messages]).
+
+    if (realmCanDeleteOwnMessageGroup != null) {
+      // realmCanDeleteOwnMessageGroup always present as of FL 291.
+      // TODO(server-10) make required, simplify, delete this FL-291 comment
```

Comment:
This is already covered by the `TODO(server-10)` on the realmCanDeleteOwnMessageGroup field itself, right? Once that's made non-nullable, the analyzer will naturally point out that this condition can be simplified.

So I think this can just apply the condition without comment.

(Similarly for the other TODO-server comments in this method.)